hilti/runtime/src/types/bytes.cc

```diff
@@ -47,34 +106,86 @@ std::tuple<bool, Bytes::const_iterator> Bytes::find(const Bytes& needle, const c
 }
 
 std::string Bytes::decode(unicode::Charset cs, unicode::DecodeErrorStrategy errors) const {
+    if ( Base::empty() )
+        return "";
+
     switch ( cs.value() ) {
         case unicode::Charset::UTF8: {
             std::string t;
 
-            auto p = reinterpret_cast<const unsigned char*>(Base::data());
-            auto e = p + Base::size();
+            auto p = Base::begin();
+            auto e = Base::end();
 
             while ( p < e ) {
-                utf8proc_int32_t cp;
-                auto n = utf8proc_iterate(p, e - p, &cp);
-
-                if ( n < 0 ) {
+                try {
+                    auto cp = utf8::next(p, e);
+                    utf8::append(cp, t);
+                } catch ( const utf8::exception& ) {
                     switch ( errors.value() ) {
                         case unicode::DecodeErrorStrategy::IGNORE: break;
-                        case unicode::DecodeErrorStrategy::REPLACE: t += "\ufffd"; break;
+                        case unicode::DecodeErrorStrategy::REPLACE: {
+                            utf8::append(unicode::REPLACEMENT_CHARACTER, t);
+                            break;
+                        }
                         case unicode::DecodeErrorStrategy::STRICT:
                             throw RuntimeError("illegal UTF8 sequence in string");
                     }
 
-                    p += 1;
-                    continue;
+                    p = std::next(p);
+                }
+            }
+
+            return t;
+        }
+
+        case unicode::Charset::UTF16BE: [[fallthrough]];
+        case unicode::Charset::UTF16LE: {
+            if ( Base::size() % 2 != 0 ) {
```

Comment:
Are there tests for this special error handling?

Reply:
I added tests to cover this. What I could not cover was the failure branches below for input `Bytes` with an even number of bytes -- while e.g., Python seems to reject certain char16 values (e.g., unpaired surrogates) the UTF16 library we use seems to normalize these to e.g., empty strings. I added a TODO for now.